### ISSUE TITLE
look into implementing a RateLimiter object

### ISSUE THREAD

**Issue by lwahlmeier** [threadly/threadly#140]
Guava has an interesting RateLimiter in its conncurrent package that could be useful for many things.  I think there implementation is a bit limited as it only lets you specify actions per second.

**Comment by jentfoo** [comment]
So threadly does have a version of a rate limiter, but it's very different from guava.  Our version is an executor, which schedules out the tasks to maintain a rate.  This does not block the thread like guava's does, though you can call in to see how many millis a future execution would block for if you wanted to make a blocking implementation.

Take a look, and let me know your thoughts:

http://threadly.github.io/threadly/javadocs/3.7.0/org/threadly/concurrent/limiter/RateLimiterExecutor.html

**Comment by jentfoo** [comment]
Oh, you can also get a future which will complete when the rate is ready to accept more tasks...that can also easily allow you to block (or you could add a listener to do an action when you go under the rate threshold)

**Comment by jentfoo** [comment]
@lwahlmeier I was thinking about it this morning....above I describe how I think our limiter is better (since it does not force you to block if you don't want to, but gives you the option to).  But I was thinking it would be trivial to make a more guava like blocking rate limiter using our existing one with a NoThreadScheduler.  An example implementation of such is here:

https://github.com/jentfoo/threadly/blob/blocking_rate_limiter/src/main/java/org/threadly/concurrent/limiter/BlockingRateLimiter.java

Do you think this should be included in threadly, or is what we provide good enough?  Maybe I should put this into threadly_examples as an example of how you can make a simple blocking rate limiter?  Maybe we are missing the mark entirely?  Let me know your thoughts.

**Comment by jentfoo** [comment]
FYI, I did realize that my provided implementation will not work with concurrent acquire calls.  So at a minimum I am not going to include this in 4.0.0, but if it is valuable we can include it in 4.1.0 or some other future release.

Let me know your thoughts @lwahlmeier if we should keep this open and add a blocking implementation, or if the non-blocking executor implementation is good enough (or better I hope).

**Comment by jentfoo** [comment]
I would like to point out I added rate limiting to ambush.  I implemented it twice, first with a blocking limiter, and second using this limiter.  I personally like threadly's way much better.

In ambush execution steps are executed on an executor, so with the blocking rate limiter we have to to block the rate at which we add things to the executor.  With our implementation we just provide a RateLimitingExecutor and it does the magic.

I am still leaving this open, I want @lwahlmeier's feedback on if it would be valuable to add a blocking rate limiter.
